File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmC4Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure
import systems.zlink.e2e.kotlin.spotservice.client.support.postAdmin
import systems.zlink.e2e.kotlin.spotservice.client.support.postJson

internal object SmC4Scenario {
    fun run() {
        val playA = Env.get("ZLINK_KOTLIN_E2E_HTTP_A_ENDPOINT")
        val gateway = Env.get("ZLINK_KOTLIN_E2E_GATEWAY_HTTP_ENDPOINT")
        val suffix = System.nanoTime().toString()
        val spotRid = "spot-sm-c4-$suffix"
        val unsubscribedSpotRid = "spot-sm-c4-unsubscribed-$suffix"
        val marker = "sm-c4-gateway"

        val created = postAdmin(playA, "/admin/create-user?rid=$spotRid")
        ensure(created.contains("\"created\":true"), "SM-C4 publish spot was not created on play-a")
        val unsubscribed = postAdmin(playA, "/admin/create-alternate?rid=$unsubscribedSpotRid")
        ensure(unsubscribed.contains("\"created\":true"), "SM-C4 unsubscribed spot was not created on play-a")

        val publish = postJson(
            gateway,
            "/spot/publish",
            mapOf("spotRid" to spotRid, "marker" to marker),
            Map::class.java
        )
        ensure(publish["scenario"] == "spot.sm-c4-publish", "SM-C4 publish operation mismatch")
        ensure(publish["rid"] == "gateway", "SM-C4 publisher was not the publish-only gateway")
        ensure(publish["spotRid"] == spotRid, "SM-C4 publish target spot mismatch")

        val snapshot = postJson(
            playA,
            "/evidence/wait",
            Contracts.EvidenceWaitReq(
                listOf("SpotMeshEvent|play-a|$spotRid|$marker"),
                10_000
            ),
            Contracts.EvidenceSnapshot::class.java
        )
        val alternateReceived = snapshot.entries.any { entry ->
            entry.marker == "SpotMeshEvent" &&
                entry.spotRid == unsubscribedSpotRid &&
                entry.value == marker
        }
        ensure(!alternateReceived, "SM-C4 unsubscribed spot received publish event")

        println("scenario SM-C4-gateway passed")
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-c4-scenario.ts
import type {
  CreateSpotReply,
  CreateSpotReq,
  SpotPublishObserveReply,
  SpotPublishReply,
  SpotPublishReq
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../Support/http-client';
import { ensure } from '../Support/scenario-assert';

export async function runSmC4(options: ClientOptions): Promise<void> {
  const spotRid = `spot-sm-c4-${Date.now()}`;
  const created = await postJson<CreateSpotReply>(options.playAUrl, '/spot/create', {
    spotRid
  } satisfies CreateSpotReq);
  ensure(
    created.spotRid === spotRid && created.nodeRid === 'play-a',
    'SM-C4 publish spot was not created on play-a.'
  );

  const unsubscribedSpotRid = `spot-sm-c4-unsubscribed-${Date.now()}`;
  const unsubscribed = await postJson<CreateSpotReply>(options.playAUrl, '/spot/create-alternate', {
    spotRid: unsubscribedSpotRid
  } satisfies CreateSpotReq);
  ensure(
    unsubscribed.spotRid === unsubscribedSpotRid && unsubscribed.nodeRid === 'play-a',
    'SM-C4 unsubscribed spot was not created on play-a.'
  );

  const marker = 'sm-c4-publish';
  const waitTask = postJson<SpotPublishObserveReply>(options.playAUrl, '/spot/publish/wait', {
    spotRid,
    marker
  } satisfies SpotPublishReq);
  const publish = await postJson<SpotPublishReply>(options.gatewayUrl, '/spot/publish', {
    spotRid,
    marker
  } satisfies SpotPublishReq);
  const observe = await waitTask;

  ensure(publish.operation === 'spot.sm-c4-publish', 'SM-C4 publish operation mismatch.');
  ensure(publish.publisherRid === 'gateway', 'SM-C4 publisher was not the publish-only gateway.');
  ensure(publish.spotRid === spotRid, 'SM-C4 publish target spot mismatch.');
  ensure(observe.operation === 'spot.sm-c4-observe', 'SM-C4 observe operation mismatch.');
  ensure(observe.received, 'SM-C4 publish-only gateway event was not received.');
  ensure(
    publish.evidence.some((line) => line.includes(`spot-publish|rid=gateway|spot=${spotRid}|marker=${marker}`)),
    'SM-C4 gateway evidence did not include publish marker.'
  );
  ensure(
    observe.evidence.some((line) => line.includes(`spot-event|rid=play-a|spot=${spotRid}|marker=${marker}`)),
    'SM-C4 evidence did not include spot event publish.'
  );
  ensure(
    observe.evidence.every((line) => !line.includes(`spot-event|rid=play-a|spot=${unsubscribedSpotRid}|marker=${marker}`)),
    'SM-C4 unsubscribed spot received publish event.'
  );

  console.log('scenario SM-C4 passed');
}
